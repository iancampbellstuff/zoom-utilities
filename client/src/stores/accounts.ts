// externals
import { defineStore } from 'pinia';
// types
import { IZoomAccountDataResponseItem } from '../../../common/src/types';

export interface IAccountsState {
    accountData: IZoomAccountDataResponseItem[];
    search: string;
}

export const useAccountsStore = defineStore({
    id: 'accounts',
    state: (): IAccountsState => ({
        accountData: [],
        search: ''
    }),
    actions: {
        resetAll() {
            this.resetAccountData();
            this.resetSearch();
        },
        resetAccountData() {
            this.accountData = [];
        },
        resetSearch() {
            this.search = '';
        },
        setAccountData(accountData: IZoomAccountDataResponseItem[]) {
            this.accountData = accountData;
        },
        setSearch(search: string) {
            this.search = search;
        }
    },
    getters: {
        getAccountData(state) {
            return state.accountData;
        },
        getSearch(state) {
            return state.search;
        }
    }
});
