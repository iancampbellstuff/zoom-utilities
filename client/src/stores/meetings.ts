import { defineStore } from 'pinia';
import { IZoomMeeting } from '../../../common/src';

export interface IMeetingsState {
    currentUserId: string;
    meetings: IZoomMeeting[];
    search: string;
    userIds: string[];
}

export const useMeetingsStore = defineStore({
    id: 'meetings',
    state: (): IMeetingsState => ({
        currentUserId: '',
        meetings: [],
        search: '',
        userIds: []
    }),
    actions: {
        resetAll() {
            this.resetCurrentUserId();
            this.resetMeetings();
            this.resetSearch();
            this.resetUserIds();
        },
        resetCurrentUserId() {
            this.currentUserId = '';
        },
        resetMeetings() {
            this.meetings = [];
        },
        resetSearch() {
            this.search = '';
        },
        resetUserIds() {
            this.userIds = [];
        },
        setCurrentUserId(currentUserId: string) {
            this.currentUserId = currentUserId;
        },
        setMeetings(meetings: IZoomMeeting[]) {
            this.meetings = meetings;
        },
        setSearch(search: string) {
            this.search = search;
        },
        setUserIds(userIds: string[]) {
            this.userIds = userIds;
        }
    },
    getters: {
        getCurrentUserId(state) {
            return state.currentUserId;
        },
        getMeetings(state) {
            return state.meetings;
        },
        getSearch(state) {
            return state.search;
        },
        getUserIds(state) {
            return state.userIds;
        }
    }
});
