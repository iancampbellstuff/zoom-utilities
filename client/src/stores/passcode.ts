import { defineStore } from 'pinia';

export interface IPasscodeState {
    firstDigit?: number;
    length?: number;
    passcode?: number;
}

export const usePasscodeStore = defineStore({
    id: 'passcode',
    state: (): IPasscodeState => ({}),
    actions: {
        resetAll() {
            this.resetFirstDigit();
            this.resetLength();
            this.resetPasscode();
        },
        resetFirstDigit() {
            delete this.firstDigit;
        },
        resetLength() {
            delete this.length;
        },
        resetPasscode() {
            delete this.passcode;
        },
        setFirstDigit(firstDigit?: number) {
            if (firstDigit) {
                this.firstDigit = firstDigit;
            }
        },
        setLength(length?: number) {
            if (length) {
                this.length = length;
            }
        },
        setPasscode(passcode?: number) {
            if (passcode) {
                this.passcode = passcode;
            }
        }
    },
    getters: {
        getFirstDigit(state) {
            return state.firstDigit;
        },
        getLength(state) {
            return state.length;
        },
        getPasscode(state) {
            return state.passcode;
        }
    }
});
