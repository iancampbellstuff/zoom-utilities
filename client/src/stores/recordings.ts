// externals
import { defineStore } from 'pinia';
// types
import { IZoomMeetingRecordingsResponseItem } from '../../../common/src/types';

export interface IRecordingsState {
    currentUserId: string;
    recordings: IZoomMeetingRecordingsResponseItem[];
    search: string;
    userIds: string[];
}

export const useRecordingsStore = defineStore({
    id: 'recordings',
    state: (): IRecordingsState => ({
        currentUserId: '',
        recordings: [],
        search: '',
        userIds: []
    }),
    actions: {
        resetAll() {
            this.resetCurrentUserId();
            this.resetRecordings();
            this.resetSearch();
            this.resetUserIds();
        },
        resetCurrentUserId() {
            this.currentUserId = '';
        },
        resetRecordings() {
            this.recordings = [];
        },
        resetSearch() {
            this.search = '';
        },
        resetUserIds() {
            this.userIds = [];
        },
        setCurrentUserId(currentUserId: string) {
            this.currentUserId = currentUserId;
        },
        setRecordings(recordings: IZoomMeetingRecordingsResponseItem[]) {
            this.recordings = recordings;
        },
        setSearch(search: string) {
            this.search = search;
        },
        setUserIds(userIds: string[]) {
            this.userIds = userIds;
        }
    },
    getters: {
        getCurrentUserId(state) {
            return state.currentUserId;
        },
        getRecordings(state) {
            return state.recordings;
        },
        getSearch(state) {
            return state.search;
        },
        getUserIds(state) {
            return state.userIds;
        }
    }
});
