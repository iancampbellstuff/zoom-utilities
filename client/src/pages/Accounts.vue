<template>
    <q-page id="accounts" class="row items-start justify-evenly">
        <q-form id="form" class="q-gutter-lg">
            <q-toolbar-title class="text-center">Accounts</q-toolbar-title>
            <q-table
                :columns="columns"
                :filter-method="(customFilter as any)"
                :filter="filter"
                :rows="data"
                outlined
                row-key="account_id"
                standout
                title="Accounts"
                v-model:pagination="paginationOptions"
            >
                <template v-slot:top>
                    <div style="width: 100%" class="row justify-between">
                        <div class="col-2">
                            <q-btn
                                @click="onRefresh"
                                color="secondary"
                                :disable="!data.length"
                                flat
                                icon="refresh"
                                id="refresh-button"
                                label="Refresh"
                                no-caps
                                wait-for-ripple
                            />
                        </div>
                        <div class="col-10">
                            <q-input
                                color="primary"
                                debounce="100"
                                v-model="search"
                                @keyup="onSearch"
                            >
                                <template v-slot:append>
                                    <q-icon name="search" />
                                </template>
                            </q-input>
                        </div>
                    </div>
                </template>
                <template v-slot:body="props">
                    <q-tr :props="props">
                        <q-td key="in_meeting" :title="props.row.in_meeting">
                            <q-icon
                                v-if="props.row.in_meeting"
                                class="status-icon"
                                color="secondary"
                                name="check"
                            />
                            <q-icon
                                v-if="!props.row.in_meeting"
                                class="status-icon"
                                color="accent"
                                name="close"
                            />
                        </q-td>
                        <q-td key="account_id" :title="props.row.account_id">{{
                            truncate(props.row.account_id)
                        }}</q-td>
                        <q-td key="topic" :title="props.row.topic">{{
                            truncate(props.row.topic)
                        }}</q-td>
                    </q-tr>
                </template>
            </q-table>
        </q-form>
    </q-page>
</template>

<style scoped lang="scss">
#accounts {
    #form {
        .q-toolbar__title {
            padding: 15px;
        }
        .q-table__container {
            width: 100%;
            .q-table__top {
                label {
                    overflow: hidden;
                }
            }
        }
        #refresh-button {
            margin-top: 8px;
            padding: 0;
        }
        .status-icon {
            width: 100%;
        }
    }
}
</style>

<script setup lang="ts">
// externals
import { useQuasar } from 'quasar';
import { computed, onBeforeMount, onMounted, ref } from 'vue';
// stores
import { useAccountsStore } from '../stores';
// types
import { IZoomAccountDataResponseItem } from '../../../common/src/types';
// utils
import { ELogLevel, getAccountData, toast } from '../utils';
import { truncate } from '../../../common/src/utils';

interface IPaginationOptions {
    rowsPerPage: number;
}

const $q = useQuasar();
const paginationOptions = ref<IPaginationOptions>({
    rowsPerPage: 10
});
const columns: any = [
    {
        align: 'center',
        field: 'in_meeting',
        label: 'Active Status',
        name: 'in_meeting',
        required: true,
        sort: (a: boolean, b: boolean) => {
            if (a === b) {
                return 0;
            } else if (a > b) {
                return -1;
            } else {
                return 1;
            }
        },
        sortable: true
    },
    {
        align: 'left',
        field: 'account_id',
        label: 'Account Name',
        name: 'account_id',
        required: true,
        sort: (a: string, b: string) => a.localeCompare(b),
        sortable: true
    },
    {
        align: 'left',
        field: 'topic',
        label: 'Active Meeting',
        name: 'topic',
        required: false,
        sort: (a?: string, b?: string) => {
            if (a && b) {
                return a.localeCompare(b);
            } else if (a) {
                return 1;
            } else {
                return -1;
            }
        },
        sortable: true
    }
];
const store = useAccountsStore();
const data = ref<IZoomAccountDataResponseItem[]>([]);
const errorOccurred = ref(false);
const search = ref('');
const filter = computed(() => {
    return search.value;
});
const customFilter = (rows: any[], searchValue: string) => {
    const filteredRows = rows.filter(
        (row: { [s: string]: unknown } | ArrayLike<unknown>) => {
            let resultFound = true;
            if (searchValue !== '') {
                const rowValues = Object.entries(row).reduce(
                    (values: string[], [key, value]) => {
                        switch (key) {
                            case 'account_id':
                            case 'topic': {
                                values.push((value as any).toString());
                                break;
                            }
                        }
                        return values;
                    },
                    []
                );
                resultFound = false;
                for (let i = 0; i < rowValues.length; i++) {
                    const rowValue = rowValues[i];
                    if (rowValue.includes(searchValue)) {
                        resultFound = true;
                        break;
                    }
                }
            }
            return resultFound;
        }
    );
    return filteredRows;
};
const showErrorMessage = () => {
    if (errorOccurred.value) {
        toast('An error occurred!', ELogLevel.ERROR);
    }
};
const onRefresh = () => {
    errorOccurred.value = false;
    $q.loading.show();
    getAccountData()
        .then((accountData: IZoomAccountDataResponseItem[]) => {
            data.value = accountData;
            store.setAccountData(accountData);
        })
        .catch(() => {
            errorOccurred.value = true;
            showErrorMessage();
        })
        .finally(() => {
            $q.loading.hide();
        });
};
const onSearch = () => {
    store.setSearch(search.value);
};

onBeforeMount(() => {
    data.value = store.accountData;
    if (!data.value?.length) {
        onRefresh();
    }
});
onMounted(() => {
    search.value = store.search;
});
</script>
