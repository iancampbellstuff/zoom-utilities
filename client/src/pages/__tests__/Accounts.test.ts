import { mount, VueWrapper } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import Quasar from 'quasar';
import { usePasscodeStore } from '../../stores';
import Accounts from '../Accounts.vue';

jest.mock('quasar', () => {
    const Notify = {
        create: jest.fn()
    };
    const useQuasar = () => ({
        loading: {
            hide: jest.fn(),
            show: jest.fn()
        }
    });
    return {
        ...jest.requireActual('quasar'),
        Notify,
        useQuasar
    };
});

describe('Accounts', () => {
    let store;
    let wrapper: VueWrapper<any>;
    beforeEach(() => {
        store = usePasscodeStore(createTestingPinia());
        wrapper = mount(Accounts, {
            global: {
                plugins: [Quasar]
            },
            store,
            stubs: [
                'q-btn',
                'q-form',
                'q-icon',
                'q-input',
                'q-page',
                'q-table',
                'q-td',
                'q-toolbar-title',
                'q-toolbar-title',
                'q-tr'
            ]
        });
    });
    test('is a Vue instance', () => {
        expect(wrapper.vm).toBeTruthy();
    });
    // TODO: "/* istanbul ignore next */" isn't working in the generated snapshot
    // eslint-disable-next-line jest/no-disabled-tests
    it.skip('should render correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
    });
});
