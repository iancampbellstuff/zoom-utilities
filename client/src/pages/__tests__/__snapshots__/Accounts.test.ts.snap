// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accounts should render correctly 1`] = `
<q-page
  class="row items-start justify-evenly"
  id="accounts"
>
  <q-form
    class="q-gutter-lg"
    id="form"
  >
    <q-toolbar-title
      class="text-center"
    >
      Accounts
    </q-toolbar-title>
    <q-table
      columns="[object Object],[object Object],[object Object]"
      filter=""
      filter-method="function (rows, searchValue) {
      /* istanbul ignore next */
      cov_1ihn90xftc().f[5]++;
      var filteredRows =
      /* istanbul ignore next */
      (cov_1ihn90xftc().s[31]++, rows.filter(function (row) {
        /* istanbul ignore next */
        cov_1ihn90xftc().f[6]++;
        var resultFound =
        /* istanbul ignore next */
        (cov_1ihn90xftc().s[32]++, true);
        /* istanbul ignore next */
        cov_1ihn90xftc().s[33]++;
        if (searchValue !== '') {
          /* istanbul ignore next */
          cov_1ihn90xftc().b[5][0]++;
          var rowValues =
          /* istanbul ignore next */
          (cov_1ihn90xftc().s[34]++, Object.entries(row).reduce(function (values, _a) {
            /* istanbul ignore next */
            cov_1ihn90xftc().f[7]++;
            var key =
              /* istanbul ignore next */
              (cov_1ihn90xftc().s[35]++, _a[0]),
              value =
              /* istanbul ignore next */
              (cov_1ihn90xftc().s[36]++, _a[1]);
            /* istanbul ignore next */
            cov_1ihn90xftc().s[37]++;
            switch (key) {
              case 'account_id':
                /* istanbul ignore next */
                cov_1ihn90xftc().b[6][0]++;
              case 'topic':
                /* istanbul ignore next */
                cov_1ihn90xftc().b[6][1]++;
                {
                  /* istanbul ignore next */
                  cov_1ihn90xftc().s[38]++;
                  values.push(value.toString());
                  /* istanbul ignore next */
                  cov_1ihn90xftc().s[39]++;
                  break;
                }
            }
            /* istanbul ignore next */
            cov_1ihn90xftc().s[40]++;
            return values;
          }, []));
          /* istanbul ignore next */
          cov_1ihn90xftc().s[41]++;
          resultFound = false;
          /* istanbul ignore next */
          cov_1ihn90xftc().s[42]++;
          for (var i =
          /* istanbul ignore next */
          (cov_1ihn90xftc().s[43]++, 0); i < rowValues.length; i++) {
            var rowValue =
            /* istanbul ignore next */
            (cov_1ihn90xftc().s[44]++, rowValues[i]);
            /* istanbul ignore next */
            cov_1ihn90xftc().s[45]++;
            if (rowValue.includes(searchValue)) {
              /* istanbul ignore next */
              cov_1ihn90xftc().b[7][0]++;
              cov_1ihn90xftc().s[46]++;
              resultFound = true;
              /* istanbul ignore next */
              cov_1ihn90xftc().s[47]++;
              break;
            } else
            /* istanbul ignore next */
            {
              cov_1ihn90xftc().b[7][1]++;
            }
          }
        } else
        /* istanbul ignore next */
        {
          cov_1ihn90xftc().b[5][1]++;
        }
        cov_1ihn90xftc().s[48]++;
        return resultFound;
      }));
      /* istanbul ignore next */
      cov_1ihn90xftc().s[49]++;
      return filteredRows;
    }"
      outlined=""
      pagination="[object Object]"
      row-key="account_id"
      rows=""
      standout=""
      title="Accounts"
    />
  </q-form>
</q-page>
`;
